gordon_janitor_gcp/gce_instance_authority.py

```diff
@@ -0,0 +1,145 @@
+    >>> print(msg)
+    {'zone': 'us-central1-b', 'resourceRecords': [...]}
+"""
+
+import aiohttp
+
+from gordon_janitor_gcp import auth
+from gordon_janitor_gcp import gce_clients
+from gordon_janitor_gcp import msg_generators
+
+
+class GCEInstanceAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): Plugin-specific configuration.
+        rrset_channel (asyncio.Queue): channel to receive instance data
+            messages.
+    """
+    def __init__(self, config, rrset_channel=None, **kwargs):
+        self.config = config
+        self.keyfile = self.config['keyfile']
+        self.rrset_channel = rrset_channel
+
+    async def get_all_projects(self):
+        """Get all active GCE projects.
+
+        Returns:
+            set: active GCE project IDs.
+        """
+        auth_client = auth.GoogleAuthClient(
+            self.keyfile,
+            scopes=['cloud-platform.read-only'],
+            session=self.session)
+        crm_client = gce_clients.CRMClient(auth_client, self.session)
+
+        active_projects = await crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def get_processed_projects(self):
+        """Get list of GCE project IDs with blacklisted ones filtered out.
+
+        Returns:
+            list: project IDs.
+        """
+        projects = await self.get_all_projects()
+        project_blacklist = set(self.config['project_blacklist'])
+        return sorted(projects - project_blacklist)
+
+    async def get_instance_filter(self):
+        """Get filter string used by instances.aggregatedList endpoint.
+
+        For filter string details, see api endpoint documentation:
+        https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList
+
+        Returns:
+            str: filter suitable for the 'filter' endpoint parameter.
+        """
+        return ''
+
+    async def get_instances_for_projects(self, projects):
+        """Get list of instance data for each project.
+
+        Args:
+            projects (list): list of project names.
+        Yields:
+            list: list of dicts that contain instance information.
+        """
+        tag_blacklist = self.config.get('tag_blacklist', [])
+        metadata_blacklist = self.config.get('metadata_blacklist', {})
+
+        auth_client = auth.GoogleAuthClient(
+            self.keyfile, scopes=['cloud-platform'], session=self.session)
+
+        gcp_client = gce_clients.GCPClient(
+            auth_client,
+            self.session,
+            tag_blacklist=tag_blacklist,
+            metadata_blacklist=metadata_blacklist)
+
+        instance_filter = await self.get_instance_filter()
+
+        for project in projects:
+            yield await gcp_client.list_instances(
+                project, instance_filter=instance_filter)
+
+    async def get_msg_generator(self):
```

Comment:
Any reason for this to be a public method? This is actually a question for the overall PR.